### ISSUE TITLE
Use CommonJS and an existing builder.

### ISSUE THREAD

**Issue by RangerMauve** [x3dom/x3dom#464]
CommonJS is a pretty popular format for structuring modules and libraries, and it lets you easily split stuff up into smaller modules and publish with something like [npm](https://www.npmjs.com/).

Any chance of this migrating from its own custom build step to just relying on CommonJS and stuff like [Browserify](http://browserify.org/) or [Webpack](http://webpack.github.io/)?

**Comment by drfrog666** [comment]
This is a great idea I was looking for x3dom in npm a cpl days ago

On 17 January, 2015 11:53:40 AM PST, RangerMauve <notifications@github.com> wrote:
>CommonJS is a pretty popular format for structuring modules and
>libraries, and it lets you easily split stuff up into smaller modules
>and publish with something like [npm](https://www.npmjs.com/).
>
>Any chance of this migrating from its own custom build step to just
>relying on CommonJS and stuff like [Browserify](http://browserify.org/)
>or [Webpack](http://webpack.github.io/)?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/x3dom/x3dom/issues/464

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by mlimper** [comment]
Hi, thanks for the proposal.

We have already thought about this, but came to the conclusion that it would not justify the restructuring work and would, moreover, make things too complex where they shouldn't be.

X3DOM is usually distributed as a single js file, along with a single css file - that way, it's pretty easy to update your copy of X3DOM, for instance (latest version can always be found under http://x3dom.org/release).
Of course, one could regard the different components of X3DOM as modules, which can also be loaded in addition to the core X3DOM version (or are alternatively already packaged all together with the core in x3dom-full.js). But in the case of X3DOM, the "module" system is trivial: A core component, and additional node pools that all depend on the core component. Introducing a dependency-managing module system looks a bit like over-engineering, at least to me.

So, to conclude, I guess we need to check what we would want to achieve, and what we would really improve by using additional build / delivery infrastructure, or by replacing the existing solution.

**Comment by RangerMauve** [comment]
The main benefit would be that you wouldn't have a custom build flow, and instead it would be the same tools that most advanced JS devs would be using anyways;. That would bring more attention from potential contributors and users in general. You'd also be able to leverage all of NPM, and also publish to it, which would help with development and spreading the word.

**Comment by andreasplesch** [comment]
My view as a small contributor would be that I see npm more as a tool for
javascript used on the server while x3dom is exclusively client code. While
I do not mind learning and having node, npm does seem like overkill to me
as well and more of a distraction for limited developer resources. On the
other hand for an experienced node person, it may be feasible to do the
packaging work in a reasonable amount of time, and then contribute it back
to the x3dom team for consideration. The main plus may be automatic
exposure if so desired.

On a perhaps related note, I would point to the large size of the git
repository as a perhaps more serious problem. But I have no good suggestion
on how to divide and conquer this large project. Based on x3d components ?

Let me also add that while it is great to see that documentation is
improving, a complete list of generated/supported events (eg.
viewpointChanged) would be much welcomed as well as a simple/rough flow
chart/narrative on how the code goes about its work from a mile up (what is
a drawable and how is it used in the drawableCollection ? what happens when
a frame is requested ? what is a scene and what is its function ? xml node
versus x3dom node ? )

-Andreas


On Tue, Jan 20, 2015 at 11:30 AM, RangerMauve <notifications@github.com>
wrote:

> The main benefit would be that you wouldn't have a custom build flow, and
> instead it would be the same tools that most advanced JS devs would be
> using anyways;. That would bring more attention from potential contributors
> and users in general. You'd also be able to leverage all of NPM, and also
> publish to it, which would help with development and spreading the word.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/x3dom/x3dom/issues/464#issuecomment-70683959>.
>



-- 
Andreas Plesch
39 Barbara Rd.
Waltham, MA 02453

**Comment by RangerMauve** [comment]
So, if someone hypothetically forked the repo and made it work with npm, but also work as a single bundle client-side, would this be appealing for the project?

**Comment by coderextreme** [comment]
I vote for a forked repo, if x3dom doesn't want to do it.  We many want to consider using bower, since it's client-side code.  I haven't googled bower and x3dom yet.

**Comment by mlimper** [comment]
It could be nice to have this feature, but I also agree that it might be overkill if x3dom needs to be setup this way.

So, it seems like a great way could be to have both options working, in parallel.

**Comment by RangerMauve** [comment]
Are you proposing that there be a build that targets UMD, then?

**Comment by mlimper** [comment]
Whatever system you all decide makes sense, my point is just that the old-fashioned way of downloading or even linking single js files, or a zip file, should still work.

**Comment by RangerMauve** [comment]
The easiest approach, then, would be to add a UMD wrapper which would allow for either globals, RequireJS, or CommonJS. [Here's an example](https://github.com/umdjs/umd/blob/master/commonjsStrictGlobal.js) of what that wrapper would look like.

**Comment by coderextreme** [comment]
As part of my build/install process, I added a wrapper.  I'm not sure it's
the right one, but it allowed me to require the x3dom-full.
